Stop using old_runs collection for runs queries

linrock opened the pull request:
@ppigazzini this PR only uses the `runs` collection for queries, so you can use it to see how the site is when everything is moved from `old_runs` to `runs`.

Fixes https://github.com/glinscott/fishtest/issues/569
Fixes https://github.com/glinscott/fishtest/issues/563

ppigazzini commented:
DEV updated, copied data with:
```
$ mongoexport -d fishtest_new -c old_runs | mongoimport -d fishtest_new -c runs
```

ppigazzini commented:
First benches show fast tests pages, fast green pages, not so fast ltc pages.
Better benches for tomorrow, Sandman at work here :)

linrock commented:
thanks for setting this up on dev! i have an idea for fixing the LTC page slowness. i'm guessing it's because LTC pages currently use regex queries to determine if a run is LTC. we can take care of this tomorrow

linrock commented:
@ppigazzini to get better performance, i added a numeric `tc_base` field to all documents in the `runs` collection and changed the indexes that are used.

before benching this, try adding the `tc_base` field to all runs:

```python
from pymongo import MongoClient

conn = MongoClient()
for run in conn['fishtest_new']['runs'].find():
    [tc_base, tc_inc] = run['args']['tc'].split('+')
    run['tc_base'] = float(tc_base)
    run['tc_inc'] = float(tc_inc)
```

then run `create_indexes.py` to re-create the indexes. then everything should be faster.

tomtor commented:
Nice! Getting rid of `old_runs` is a significant cleanup.

With all indices now in place it should give a good performance.

ppigazzini commented:
Unfortunately after some benches (tried several times) this PR is a slowdown

|   |  master | runs | script  |
|---|---|---|---|
| tests   | 0m31.417s | 1m9.555s | 1m7.458s |
| ltc       | 1m2.148s | 1m26.236s | 1m24.646s |
| green  | 0m10.586s | 1m21.597s | 1m17.965s |

Script updated to deal with data inconsistency:
```python
import re
from pymongo import MongoClient

conn = MongoClient()
for run in conn['fishtest_new']['runs'].find():
  tc_value = run['args']['tc']
  try:
    [tc_base, *tc_inc] = re.split('[+ /]', tc_value)
    run['tc_base'] = float(tc_base)
    if tc_inc:
      run['tc_inc'] = float(tc_inc[0])
    else:
      run['tc_inc'] = 0.0
  except Exception as e:
    print(tc_value)
    print(e)
```
output
```

could not convert string to float:
5+0.0.5
could not convert string to float: '0.0.5'
 20+0.2
could not convert string to float:
 20+0.2
could not convert string to float:
```

ppigazzini commented on fishtest/fishtest/rundb.py:
```diff
@@ -109,16 +103,19 @@ def new_run(self, base_tag, new_tag, num_games, tc, book, book_depth,
     if spsa is not None:
       run_args['spsa'] = spsa
 
+    [tc_base, tc_inc] = tc.split('+')
```
unfortunately there are tc written as "20", "20/5" and some typo/garbage as "20+0.0.5", "20+20+0.5", "inf"

linrock replied:
thanks, i've updated this logic to deal with the garbage values. we should add validations on the runs documents later to prevent garbage from getting in there

tomtor commented:
> Unfortunately after some benches (tried several times) this PR is a slowdown

But is it really too slow? Performance/elegant-code/correct-functionality is a trade off.

Ltc/green queries are rare and the main page is cached.

IMO this PR is an overall improvement.

The only real performance bottleneck is fishtest handling of over 10000 worker cores, not the info pages.

ppigazzini commented:
> But is it really too slow? Performance/elegant-code/correct-functionality is a trade off.

Browsing on DEV is a bit painful wrt PROD:
https://dfts-0.pigazzini.it/tests
https://tests.stockfishchess.org/tests

LTC pages is broken:
https://dfts-0.pigazzini.it/tests?ltc_only=1

tomtor commented:
> Browsing on DEV is a bit painful wrt PROD:

Yes but 16000 cores also give a slow down?

ppigazzini commented:
> Yes but 16000 cores also give a slow down?

Browsing on PROD w/ 10k cores and w/ old_runs is fast (tried this morning) 
Browsing on DEV w/ 0 cores and w/o old_runs is slow

ppigazzini commented:
@tomtor I ran `scavenge.py` by hand, now DEV has 0 cores, browsing is always slow

linrock commented:
thanks for testing! i'll generate a bunch of test data on my dev server and see if i can reproduce and fix the slowdown issues

linrock commented:
@ppigazzini i can't reproduce the slowness issues. with over 100,000 runs in my test db on a $5 VPS with 1GB ram, the runs pages are all very fast for me. i looked at the query plans for the runs queries and they're all `IXSCAN -> FETCH -> LIMIT`. that's the fastest possible query plan.

1. did you re-create the indexes with `create_indexes.py` from this branch?
2. if the LTC page had zero results, that means the `tc_base` field migration failed.

---
try this for a way to gracefully handle bad data in the `tc_base` field migration:
```python
import re
from pymongo import MongoClient

conn = MongoClient()
for run in conn['fishtest_new']['runs'].find():
    tc_base = re.search('^(\d+)', run['args']['tc'])
    if tc_base:
        tc_base = float(tc_base[1])
    run['tc_base'] = tc_base
```

---
to double-check that the indexes were created, try running this script:
```python
from pprint import pprint
from pymongo import MongoClient, DESCENDING

conn = MongoClient()
db = conn['fishtest_new']

def print_query_plan(mongo_cursor):
    explanation = mongo_cursor.explain()
    winning_plan = explanation['queryPlanner']['winningPlan'];
    if winning_plan:
        pprint(winning_plan) # Query starts with an index scan
    else:
        pprint(explanation['queryPlanner']) # Slow query - collection scan

def explain_runs_queries():
    runs_col = db['runs']

    print("runs - index information")
    pprint(runs_col.index_information())
    print("")

    # homepage
    print("-- Homepage - get_finished_runs() --")
    print_query_plan(runs_col.find({
      'finished': True,
    }, limit=50, sort=[('last_updated', DESCENDING)]))
    print("")

    # username links
    print("-- Username - get_finished_runs() --")
    print_query_plan(runs_col.find({
      'finished': True,
      'args.username': 'userA'
    }, limit=50, sort=[('last_updated', DESCENDING)]))
    print("")

    # LTC (Slow) - get_finished_runs()
    print("-- LTC get_finished_runs() --")
    print_query_plan(runs_col.find({
      'finished': True,
      'tc_base': { '$gte': 40 }
    }, limit=50, sort=[('last_updated', DESCENDING)]))
    print("")

    # Greens - get_finished_runs()
    print("-- Greens get_finished_runs() --")
    print_query_plan(runs_col.find({
      'finished': True,
      'results_info.style': '#44EB44'
    }, limit=50, sort=[('last_updated', DESCENDING)]))
    print("")

explain_runs_queries()
```

if the index was correctly created, you'll see that all queries use `IXSCAN -> FETCH -> LIMIT`

vondele commented on fishtest/fishtest/rundb.py:
```diff
@@ -109,16 +103,20 @@ def new_run(self, base_tag, new_tag, num_games, tc, book, book_depth,
     if spsa is not None:
       run_args['spsa'] = spsa
 
+    tc_base = re.search('^(\d+)', tc)
+    if tc_base:
+      tc_base = float(tc_base[1])
```
`re.search('^(\d+)', tc)` will this match also non-integer base tc ? (like "13.5+0.1") 
It might not really matter right now, but maybe it is confusing later.

linrock replied:
good catch, i updated the regex to match `13.5` and such. it won't work for `.5` and probably other edge cases for now

ppigazzini commented:
I will try to redo the process with your latest commit.

Anyway I reloaded the db with mongorestore.
Here is the output of my test script:
```bash
$ ${VENV}/bin/python x/show_indexes.py

Current Indexes:
{'_id_': {'key': [('_id', 1)], 'ns': 'fishtest_new.actions', 'v': 2},
 '_id_-1_action_1_username_1': {'key': [('_id', -1),
                                        ('action', 1),
                                        ('username', 1)],
                                'ns': 'fishtest_new.actions',
                                'v': 2}}
{'_id_': {'key': [('_id', 1)], 'ns': 'fishtest_new.clop', 'v': 2}}
{'_id_': {'key': [('_id', 1)], 'ns': 'fishtest_new.deltas', 'v': 2}}
{'_id_': {'key': [('_id', 1)], 'ns': 'fishtest_new.flag_cache', 'v': 2}}
{'_id_': {'key': [('_id', 1)], 'ns': 'fishtest_new.old_runs', 'v': 2},
 'finished_-1_args.username_1_args.tc_1_results_info.style_1__id_-1': {'key': [('finished',
                                                                                -1),
                                                                               ('args.username',
                                                                                1),
                                                                               ('args.tc',
                                                                                1),
                                                                               ('results_info.style',
                                                                                1),
                                                                               ('_id',
                                                                                -1)],
                                                                       'ns': 'fishtest_new.old_runs',
                                                                       'v': 2}}
{'_id_': {'key': [('_id', 1)], 'ns': 'fishtest_new.old_user_cache', 'v': 2}}
{'_id_': {'key': [('_id', 1)], 'ns': 'fishtest_new.pgns', 'v': 2},
 'run_id_-1': {'key': [('run_id', -1)], 'ns': 'fishtest_new.pgns', 'v': 2}}
{'_id_': {'key': [('_id', 1)], 'ns': 'fishtest_new.regression_data', 'v': 2}}
{'_id_': {'key': [('_id', 1)], 'ns': 'fishtest_new.runs', 'v': 2},
 'last_updated_-1': {'key': [('last_updated', -1)],
                     'ns': 'fishtest_new.runs',
                     'partialFilterExpression': SON([('finished', False)]),
                     'v': 2},
 'last_updated_-1_results_info.style_1_tc_base_-1_args.username_1': {'key': [('last_updated',
                                                                              -1),
                                                                             ('results_info.style',
                                                                              1),
                                                                             ('tc_base',
                                                                              -1),
                                                                             ('args.username',
                                                                              1)],
                                                                     'ns': 'fishtest_new.runs',
                                                                     'partialFilterExpression': {'finished': True},
                                                                     'v': 2}}
{'_id_': {'key': [('_id', 1)], 'ns': 'fishtest_new.top_month', 'v': 2}}
{'_id_': {'key': [('_id', 1)], 'ns': 'fishtest_new.user_cache', 'v': 2}}
{'_id_': {'key': [('_id', 1)], 'ns': 'fishtest_new.users', 'v': 2},
 'username_1': {'key': [('username', 1)],
                'ns': 'fishtest_new.users',
                'unique': True,
                'v': 2}}
```
Here is the output of your test script:
```bash
${VENV}/bin/python x/show_indexes2.py
runs - index information
{'_id_': {'key': [('_id', 1)], 'ns': 'fishtest_new.runs', 'v': 2},
 'last_updated_-1': {'key': [('last_updated', -1)],
                     'ns': 'fishtest_new.runs',
                     'partialFilterExpression': SON([('finished', False)]),
                     'v': 2},
 'last_updated_-1_results_info.style_1_tc_base_-1_args.username_1': {'key': [('last_updated',
                                                                              -1),
                                                                             ('results_info.style',
                                                                              1),
                                                                             ('tc_base',
                                                                              -1),
                                                                             ('args.username',
                                                                              1)],
                                                                     'ns': 'fishtest_new.runs',
                                                                     'partialFilterExpression': {'finished': True},
                                                                     'v': 2}}

-- Homepage - get_finished_runs() --
{'inputStage': {'filter': {'finished': {'$eq': True}},
                'inputStage': {'direction': 'forward',
                               'indexBounds': {'args.username': ['[MinKey, '
                                                                 'MaxKey]'],
                                               'last_updated': ['[MaxKey, '
                                                                'MinKey]'],
                                               'results_info.style': ['[MinKey, '
                                                                      'MaxKey]'],
                                               'tc_base': ['[MaxKey, MinKey]']},
                               'indexName': 'last_updated_-1_results_info.style_1_tc_base_-1_args.username_1',
                               'indexVersion': 2,
                               'isMultiKey': False,
                               'isPartial': True,
                               'isSparse': False,
                               'isUnique': False,
                               'keyPattern': {'args.username': 1,
                                              'last_updated': -1,
                                              'results_info.style': 1,
                                              'tc_base': -1},
                               'multiKeyPaths': {'args.username': [],
                                                 'last_updated': [],
                                                 'results_info.style': [],
                                                 'tc_base': []},
                               'stage': 'IXSCAN'},
                'stage': 'FETCH'},
 'limitAmount': 50,
 'stage': 'LIMIT'}

-- Username - get_finished_runs() --
{'inputStage': {'filter': {'$and': [{'args.username': {'$eq': 'userA'}},
                                    {'finished': {'$eq': True}}]},
                'inputStage': {'direction': 'forward',
                               'indexBounds': {'args.username': ['[MinKey, '
                                                                 'MaxKey]'],
                                               'last_updated': ['[MaxKey, '
                                                                'MinKey]'],
                                               'results_info.style': ['[MinKey, '
                                                                      'MaxKey]'],
                                               'tc_base': ['[MaxKey, MinKey]']},
                               'indexName': 'last_updated_-1_results_info.style_1_tc_base_-1_args.username_1',
                               'indexVersion': 2,
                               'isMultiKey': False,
                               'isPartial': True,
                               'isSparse': False,
                               'isUnique': False,
                               'keyPattern': {'args.username': 1,
                                              'last_updated': -1,
                                              'results_info.style': 1,
                                              'tc_base': -1},
                               'multiKeyPaths': {'args.username': [],
                                                 'last_updated': [],
                                                 'results_info.style': [],
                                                 'tc_base': []},
                               'stage': 'IXSCAN'},
                'stage': 'FETCH'},
 'limitAmount': 50,
 'stage': 'LIMIT'}

-- LTC get_finished_runs() --
{'inputStage': {'filter': {'$and': [{'finished': {'$eq': True}},
                                    {'tc_base': {'$gte': 40}}]},
                'inputStage': {'direction': 'forward',
                               'indexBounds': {'args.username': ['[MinKey, '
                                                                 'MaxKey]'],
                                               'last_updated': ['[MaxKey, '
                                                                'MinKey]'],
                                               'results_info.style': ['[MinKey, '
                                                                      'MaxKey]'],
                                               'tc_base': ['[MaxKey, MinKey]']},
                               'indexName': 'last_updated_-1_results_info.style_1_tc_base_-1_args.username_1',
                               'indexVersion': 2,
                               'isMultiKey': False,
                               'isPartial': True,
                               'isSparse': False,
                               'isUnique': False,
                               'keyPattern': {'args.username': 1,
                                              'last_updated': -1,
                                              'results_info.style': 1,
                                              'tc_base': -1},
                               'multiKeyPaths': {'args.username': [],
                                                 'last_updated': [],
                                                 'results_info.style': [],
                                                 'tc_base': []},
                               'stage': 'IXSCAN'},
                'stage': 'FETCH'},
 'limitAmount': 50,
 'stage': 'LIMIT'}

-- Greens get_finished_runs() --
{'inputStage': {'filter': {'$and': [{'finished': {'$eq': True}},
                                    {'results_info.style': {'$eq': '#44EB44'}}]},
                'inputStage': {'direction': 'forward',
                               'indexBounds': {'args.username': ['[MinKey, '
                                                                 'MaxKey]'],
                                               'last_updated': ['[MaxKey, '
                                                                'MinKey]'],
                                               'results_info.style': ['[MinKey, '
                                                                      'MaxKey]'],
                                               'tc_base': ['[MaxKey, MinKey]']},
                               'indexName': 'last_updated_-1_results_info.style_1_tc_base_-1_args.username_1',
                               'indexVersion': 2,
                               'isMultiKey': False,
                               'isPartial': True,
                               'isSparse': False,
                               'isUnique': False,
                               'keyPattern': {'args.username': 1,
                                              'last_updated': -1,
                                              'results_info.style': 1,
                                              'tc_base': -1},
                               'multiKeyPaths': {'args.username': [],
                                                 'last_updated': [],
                                                 'results_info.style': [],
                                                 'tc_base': []},
                               'stage': 'IXSCAN'},
                'stage': 'FETCH'},
 'limitAmount': 50,
 'stage': 'LIMIT'}
```

linrock commented:
sounds good, the output from your two scripts look good

ppigazzini commented:
@linrock done the work from zero (mongorestore from PROD, mongoexport | mongoimport, set_tc_base.py, create_indexes.py), here is your script updated to latest commit:
```python
import re
from pymongo import MongoClient

conn = MongoClient()
for run in conn['fishtest_new']['runs'].find():
    tc_base = re.search('^(\d+(\.\d+)?)', run['args']['tc'])
    if tc_base:
        tc_base = float(tc_base[1])
    run['tc_base'] = tc_base
```

Browsing on DEV is slow wrt PROD:
https://dfts-0.pigazzini.it/tests
https://tests.stockfishchess.org/tests

LTC page is always broken:
https://dfts-0.pigazzini.it/tests?ltc_only=1

Bench numbers are confirmed.

linrock commented:
my bad, i realized the migration snippet (which i wasn't using) wasn't saving the results.. try this:
```python
import re
from pymongo import MongoClient

conn = MongoClient()
runs = conn['fishtest_new']['runs']
for run in runs.find():
    tc_base = re.search('^(\d+(\.\d+)?)', run['args']['tc'])
    if tc_base:
        tc_base = float(tc_base[1])
    run['tc_base'] = tc_base
    runs.save(run)
```

then runs will show up in the LTC page.

i'll think of a way to debug the slowness issue and let you know.

ppigazzini commented:
@linrock give me 30 minutes, I'm testing CSRF now

linrock commented:
@ppigazzini a way to debug slow mongo queries in dev is this:

in `/etc/mongod.conf` set the verbosity of commands and queries to 3:
```
systemLog:
  destination: file
  logAppend: true
  path: /var/log/mongodb/mongod.log
  component:
    command:
      verbosity: 3
    query:
      verbosity: 3
```

after restarting mongo, you can watch for query speeds by tailing the logs:

`tail -f /var/log/mongodb/mongod.log | grep -E "[0-9]+ms"`

then if you browse the site, you'll see the query timings. if any query seems really slow (over 500ms), let me know and i can investigate

linrock commented:
@ppigazzini i found and fixed a few performance issues in this PR. i believe a big problem was that count queries could be extremely slow for large collections if the working data set didn't fit in RAM. that's fixed now by using fast estimated counts when possible with multiple partial indexes.

the new data migration script (untested):

```python
import re
from pymongo import MongoClient

conn = MongoClient()
runs = conn['fishtest_new']['runs']
for run in runs.find():
    tc_base = re.search('^(\d+(\.\d+)?)', run['args']['tc'])
    if tc_base:
        tc_base = float(tc_base[1])
    run['tc_base'] = tc_base
    if run['results_info'].get('style') == '#44EB44':
        run['is_green'] = True
    runs.save(run)
```

ppigazzini commented:
@linrock OK, I'm starting to test on DEV.

- PROD and DEV same ISP but in two different zones
- PROD 5GB RAM + 0GB swap (no swap), HD fast
```
time (dd if=/dev/zero of=/tmp/test bs=64k count=64k > /dev/null; sync)
65536+0 records in
65536+0 records out
4294967296 bytes (4.3 GB, 4.0 GiB) copied, 14.3773 s, 299 MB/s
real    0m14.503s
```

- DEV 6GB RAM + 3GB swap, HD slower (yes, I should build a new DEV...)
```
time (dd if=/dev/zero of=/tmp/test bs=64k count=64k > /dev/null; sync)
65536+0 records in
65536+0 records out
4294967296 bytes (4.3 GB, 4.0 GiB) copied, 49.7722 s, 86.3 MB/s
real    0m50.026s
```

tomtor commented:
@linrock No issue for me but

https://dfts-0.pigazzini.it/tests/user/Lolligerhans

shows as count all tests (77210), just as master.

linrock commented:
@ppigazzini just to confirm, did you set the `vm.swappiness` to a low value on the dev machine? by default, linux systems will swap a lot and that would destroy MongoDB performance even if the dataset could otherwise fit entirely in RAM.

mongodb recommends setting `vm.swappiness` to 0 or 1:
https://docs.mongodb.com/manual/administration/production-notes/#set-vm-swappiness-to-1-or-0

@tomtor i'm guessing @ppigazzini loaded up master on dev before you went to the that link, since in this branch those counts should be fixed.

ppigazzini commented:
@tomtor DEV is testing #581 at the moment. I confirm that with #574 LTC and Green have the correct tests count.
@linrock I will try to bench `vm.swappiness`

ppigazzini commented:
@linrock DEV and PROD are OpenVZ containers, swap is usually in RAM and it's used only if the RAM is exhausted, see https://docs.openvz.org/openvz_users_guide.webhelp/_configuring_main_vswap_parameters.html

Here is DEV during a bench:

```
top - 22:09:58 up 5 days, 22:38,  1 user,  load average: 3.86, 1.28, 0.79
Tasks:  30 total,   1 running,  29 sleeping,   0 stopped,   0 zombie
%Cpu0  :  76.7/8.3    85[|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||               ]
%Cpu1  :  63.7/6.3    70[||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||                              ]
%Cpu2  :  49.3/4.0    53[|||||||||||||||||||||||||||||||||||||||||||||||||||||                                               ]
%Cpu3  :  79.7/9.3    89[|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||           ]
GiB Mem : 38.4/6.000    [||||||||||||||||||||||||||||||||||||||                                                              ]
GiB Swap:  0.5/3.000    [|                                                                                                   ]

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
    1 root      20      159348   3432   2108 S        0.1   0:03.62 init -z
    2 root      20                           S                       `- [kthreadd/13518]
    3 root      20                           S                           `- [khelper]
   56 root      20      114424  11240   8356 S        0.2   4:28.90  `- /lib/systemd/systemd-journald
  166 root      20       42092    828    784 S        0.0   0:00.40  `- /lib/systemd/systemd-udevd
  169 systemd+  20       71848    856    800 S        0.0   0:00.48  `- /lib/systemd/systemd-networkd
  181 root      20       28344    996    812 S        0.0   0:01.19  `- /usr/sbin/cron -f
  183 root      20       62116   1304   1152 S        0.0   0:00.65  `- /lib/systemd/systemd-logind
  184 message+  20       47772   1800   1356 S        0.0   0:00.39  `- /usr/bin/dbus-daemon --system --address=systemd: --nofork --nopidfile --systemd-activation --syslog-only
  186 syslog    20      189016   2056    744 S        0.0   2:43.29  `- /usr/sbin/rsyslogd -n
  327 root      20      184300   3788   1512 S        0.1   0:00.10  `- /usr/bin/python3 /usr/share/unattended-upgrades/unattended-upgrade-shutdown --wait-for-signal
  386 root      20       24176    552    548 S        0.0            `- /usr/sbin/xinetd -pidfile /run/xinetd.pid -stayalive -inetd_compat -inetd_ipv6
  586 root      20       72288    972    916 S        0.0            `- /usr/sbin/sshd -D
25261 root      20       97180   1524   1148 S        0.0   0:00.08      `- sshd: usr00 [priv]
25277 usr00     20       97180    844    420 S        0.0   0:10.21          `- sshd: usr00@pts/1
25278 usr00     20       18684   1884   1276 S        0.0   0:00.18              `- -bash
12851 usr00     20       36620   1828   1280 R   0.3  0.0   0:00.42                  `- top
  592 root      20      100972    348    324 S        0.0            `- /usr/sbin/saslauthd -a pam -c -m /var/run/saslauthd -n 2
  593 root      20      100972     24        S        0.0                `- /usr/sbin/saslauthd -a pam -c -m /var/run/saslauthd -n 2
  594 Debian-+  20       61256    540    392 S        0.0   0:00.10  `- /usr/sbin/exim4 -bd -q30m
  602 root      20      142260   1620     28 S        0.0            `- nginx: master process /usr/sbin/nginx -g daemon on; master_process on;
  603 www-data  20      156860  16488   3168 S   1.0  0.3  26:02.57      `- nginx: worker process
  604 www-data  20      146640   7144   2520 S   0.3  0.1   0:38.35      `- nginx: worker process
  605 www-data  20      143136   3784   1884 S        0.1   0:00.43      `- nginx: worker process
  606 www-data  20      142596   2588    952 S        0.0   0:00.02      `- nginx: worker process
26393 mongodb   20     3212652 1.610g   7036 S 286.3 26.8 119:53.21  `- /usr/bin/mongod --config /etc/mongod.conf
12537 root      20       13008    848    712 S        0.0            `- /sbin/agetty -o -p -- \u --noclear tty2 linux
12538 root      20       13008    848    716 S        0.0            `- /sbin/agetty -o -p -- \u --noclear --keep-baud console 115200,38400,9600 linux
12703 usr00     20     2032068 627256  13144 S  15.0 10.0   0:42.38  `- /home/usr00/fishtest/fishtest/env/bin/python3 /home/usr00/fishtest/fishtest/env/bin/pserve production.ini http_port=6543
12704 usr00     20     1799892  72356  12936 S        1.2   0:02.01  `- /home/usr00/fishtest/fishtest/env/bin/python3 /home/usr00/fishtest/fishtest/env/bin/pserve production.ini http_port=6544

```

ppigazzini commented:
@linrock [DEV](https://dfts-0.pigazzini.it/tests) updated and browsing is snappy, from benches the PR looks good :)
|   |  master | pr |
|---|---|---|
| tests   | 0m27.724s | 0m36.315s |
| ltc       | 1m1.288s | 0m52.694s |
| green  | 0m11.508s | 0m47.278s |

I updated your script to deal with some data inconsistency:
```python
import re
import sys
from pymongo import MongoClient

conn = MongoClient()
runs = conn['fishtest_new']['runs']
for run in runs.find():
    tc_base = re.search('^(\d+(\.\d+)?)', run['args']['tc'])
    if tc_base:
        tc_base = float(tc_base[1])
    run['tc_base'] = tc_base
    try:
        if run['results_info'].get('style') == '#44EB44':
            run['is_green'] = True
    except Exception as e:
        print("Unexpected error:", sys.exc_info()[0], e)
    runs.save(run)
```

linrock commented:
nice! i looked through dev and most of the tests pages (homepage, greens, ltc) feel relatively fast.

tomtor commented on fishtest/tests/test_rundb.py:
```diff
@@ -9,6 +11,11 @@ class CreateRunDBTest(unittest.TestCase):
 
   def setUp(self):
     self.rundb = util.get_rundb()
+    self.rundb.runs.ensure_index(
```
Just a question: why defined both here and in `create_indices.py`?

linrock replied:
this index is on the test db `fishtest_test` which `create_indexes.py` doesn't yet touch (maybe it should). using a separate db is to prevent unexpected behavior for happening in case someone runs the test script from a dev/prod environment.

this `ensure_index` call is needed in this test because we're explicitly using this `finished_ltc_runs` index in `get_finished_runs()` and without ensuring the index here that call would fail

tomtor replied:
Ah, I missed that it was in the testing code. Sorry for the noise.

tomtor commented:
@linrock Great job!

linrock commented:
@ppigazzini fyi, i saw the comment about slow actions when the searched username doesn't exist https://github.com/glinscott/fishtest/issues/553#issuecomment-613259607 when using the indexes from this branch.

i updated the actions collection indexes here to fix that performance problem: https://github.com/glinscott/fishtest/pull/574/commits/c55629829ca239672af78c3ba64a6179ef2e0808

ppigazzini commented:
@linrock please rebase :)

linrock commented:
@ppigazzini rebased, merge conflict fixed, and squashed :)

ppigazzini commented:
@linrock please rebase :(
I'll prepare a script to automate the switch on PROD (take backup, avoid errors, reduce the offline time)

ppigazzini commented:
@linrock thank you, started the final test on DEV

linrock commented:
@ppigazzini no prob!

ppigazzini commented:
@noobpwnftw started the update to merge this PR, PROD will be off line for ~ 30 minutes

ppigazzini commented:
@linrock @tomtor @vondele please review PROD after pulling the PR and updating the DB

```
real    114m26.898s
user    63m33.254s
sys     4m37.059s
```

linrock commented:
prod seems ok to me so far. the larger page numbers in the runs pages are a bit slow because the mongo queries are using `skip` but that was happening before too.

ppigazzini commented:
PR merged and PROD updated.
Great job @linrock :)

linrock commented:
@ppigazzini thanks :) also just to confirm, cron was updated to stop running `move2old_runs.py` right?

ppigazzini commented:
@linrock I commented out `move2old_runs.p` before starting the script. I stopped cron to stop the other maintenance jobs: `delta_update_users.py`, `scavenge.py`

linrock commented:
cool, safe to remove that script from the repo and update the wiki then:
https://github.com/glinscott/fishtest/wiki/Fishtest-server-setup

i can open a PR unless you want to remove it. i'll update the wiki

vondele commented:
PROD seems fine to me.
